Unwrap Qt 5 packages

ttuegel opened the pull request:
### Motivation

The Qt 5 wrappers are difficult to maintain and prevent runtime interoperability between packages. The latter is an important goal for desktop environments such as KDE. Fortunately, the wrappers are also unnecessary.

This pull request removes `makeQtWrapper` and `kdeWrapper`. Instead of wrapping programs, required runtime dependencies are propagated through `propagatedUserEnvPkgs`. However, this does mean that some applications may not work correctly if invoked directly from the Nix store. This necessitates other changes, most notably: Qt plugins are now installed in a different prefix for each minor version to prevent conflicts.

### Testing

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I have also run the relevant NixOS tests.

---

vandenoever committed:
gpgme: enable Qt 5 bindings

ttuegel committed:
breeze-gtk: bump GTK3 version

ttuegel committed:
breeze-gtk: patch builder to find GTK2 pixmap engine

This does not seem to be strictly necessary, but it also cannot hurt.

ttuegel committed:
qgpgme: init at 1.9.0

ttuegel committed:
akonadi: fix runtime paths

ttuegel committed:
qt5: use distinct `lib/qt-5.x` prefix for each minor version

Using a distinct prefix for plugins and QML libraries allows multiple Qt 5 minor
versions to coexist in the same environment.

ttuegel committed:
plasma5: rename plasmaPackage to mkDerivation

ttuegel committed:
kdeFrameworks: rename kdeFramework to mkDerivation

ttuegel committed:
Remove kdeWrapper

ttuegel committed:
sddm: take themes from system environment

ttuegel committed:
qmake: unify Darwin and Linux setup hooks

ttuegel committed:
qt5: remove makeQtWrapper

ttuegel committed:
qtbase: setup required environment in GTK3 platform theme plugin

ttuegel committed:
qt5: separate plugins from libraries

ttuegel committed:
akonadi: do not store path to mysqld in configuration

ttuegel committed:
akonadi: do not store path to pg_ctl in configuration

ttuegel committed:
qt5: use one output by default

ttuegel committed:
qtbase: fix plugin path in CMake modules

ttuegel committed:
startkde: use absolute paths to required programs

The dependency on socat is removed because kwallet setup is deferred to kwallet-pam.

ttuegel committed:
qttools: move runtime tools in postFixup

ttuegel committed:
prison: init at 5.33.0

ttuegel committed:
kdeFrameworks: fixup inputs and outputs

- Reduce environment pollution with a separate $bin output containing programs,
  plugins, and shared data. Libraries remain in $out and are not installed into
  the environment.
- Only propagate build inputs as required.

ttuegel committed:
libkscreen: use QLatin1String instead of QStringLiteral

QStringLiteral stores strings in UTF-16 format, preventing Nix from detecting
dependencies.

ttuegel committed:
kwindowsystem: use QLatin1String instead of QStringLiteral

ttuegel committed:
qt58.qtbase: use QLatin1String instead of QStringLiteral

ttuegel committed:
qt58.qtserialport: use QLatin1String instead of QStringLiteral

ttuegel committed:
qt58.qtwebkit: use QLatin1String instead of QStringLiteral

ttuegel committed:
qt56.qtwebkit: use QLatin1String instead of QStringLiteral

ttuegel committed:
qt56.qtserialport: use QLatin1String instead of QStringLiteral

ttuegel committed:
kdeFrameworks: sort according to tier

ttuegel committed:
kio: print plugin name when loading fails

ttuegel committed:
kcmutils: print plugin name when loading fails

ttuegel committed:
plasma-desktop: pass path to hwclock through preprocessor macro

ttuegel committed:
kwin: pass path to Xwayland through preprocessor macro

ttuegel committed:
plasma5: fixup inputs and outputs

ttuegel committed:
kwallet-pam: set path to socat

ttuegel committed:
kcmutils: fix kcm plugin path

ttuegel committed:
kservice: propagate shared_mime_info to environment

ttuegel committed:
nixpkgs: remove comments from manual about wrapping Qt packages

It is no longer necessary to wrap Qt packages

ttuegel committed:
nixpkgs: remark about running Qt applications

ttuegel committed:
qt5: rename qmakeHook to qmake

ttuegel committed:
qt5: move kdeDerivation to qt5.mkDerivation

ttuegel committed:
kdeApplications: rename kdeApp to mkDerivation

FRidh commented:
> The Qt 5 wrappers are difficult to maintain and prevent runtime interoperability between packages.

Could you refer to examples? Can `nix-shell` (or a future `nix run`) deal with `propagatedUserEnvPkgs`?

ttuegel commented:
> Could you refer to examples?

#### Maintainability

Whenever an old package gets a new dependency, I have to go through the whole dependency graph and prune the `$dev` outputs (the wrappers tend to be a little zealous about retaining outputs).

#### Interoperability

In another branch, I am packaging the KDE PIM suite (KMail, KOrganizer, Kontact, etc.). All the KDE packages are designed to detect optional runtime dependencies, but the PIM suite takes this to an extreme. With the wrappers, we either need to package the PIM suite as one monolith or tolerate large duplicate symlink farms for each package.

I don't like the monolith approach because it requires that we fight upstream's design for their software. It is designed to be installed into an environment and cooperate with other packages in that environment, and I don't like subverting that goal. Also, I imagine many users do not need or want the entire PIM suite; for example, they may only use KMail and do not want to install gigabytes of unrelated software.

We had a solution in `kdeWrapper` for dealing with large symlink farms. This works fairly well, but only if you don't use `nix-env`. I have some personal feelings about `nix-env` (I don't like it) but it is part of the Nix ecosystem and we need it to work right. Ultimately, `kdeWrapper` creates an environment to exist alongside the environments created by `nix-env` or `nixos-rebuild`. It duplicates the functionality of those environments, adding another layer of complexity, and it breaks `nix-env`... I think these are good reasons to seek another solution.


> Can nix-shell (or a future nix run) deal with propagatedUserEnvPkgs?

`nix-shell` cannot do so by itself, but I could modify the Qt tooling slightly to automatically generate an environment for each package that could be used with `nix-shell`. I think we can borrow a trick from `cabal2nix` to detect if we're running in `nix-shell` and switch to the environment automatically, instead of the package's default output.

FRidh commented:
Thanks for the explanation @ttuegel. We should indeed not fight upstream's design. I thought it be good to have the argumentation you gave now because its otherwise hard to track why certain design choices were made and abandoned.

AndersonTorres commented:
We can then eliminate any occurrence of qt?Wrappers?

2017-06-03 12:39 GMT-03:00 Frederik Rietdijk <notifications@github.com>:

> Thanks for the explanation @ttuegel <https://github.com/ttuegel>. We
> should indeed not fight upstream's design. I thought it be good to have the
> argumentation you gave now because its otherwise hard to track why certain
> design choices were made and abandoned.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/26336#issuecomment-305982601>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFrc9iFRWY5dhFviCAbndeT_J-IK4PtRks5sAX5LgaJpZM4Nu4D2>
> .
>

olejorgenb commented:
>  However, this does mean that some applications may not work correctly if invoked directly from the Nix store.

What does that mean exactly? Some apps must be launched through a .desktop file? Simply typing the command name in a shell wont work?

ttuegel commented:
> What does that mean exactly? Some apps must be launched through a .desktop file? Simply typing the command name in a shell wont work?

It means that `nix-build -A foo; ./result/bin/foo` may not work. `nix-env -iA foo; foo` will always work. I am taking steps to see that `nix-shell -p foo` also works.

olejorgenb commented:
Thanks for the clarification. That doesn't sound too bad. As a non-desktop-suite user I got a bit worried there for a second :)

A bit off topic, but just in case these changes make it simpler/possible to fix the icon issue in kde apps installed/run as standalone apps: 

Okular5 for instance works when installed standalone (lots of icons missing, but not that big a deal actually). BUT - startup time is affected due to how icon lookup works: okular makes 400k+ stat calls looking for icons. 

This seems to be because `QT_QPA_PLATFORMTHEME` isn't set. Setting it to eg. "kde" reduces the stat calls to 9k and startup time is at least cut in half on my system.

Mic92 commented on doc/languages-frameworks/qt.xml:
```diff
@@ -20,13 +20,13 @@
 <section xml:id="ssec-qt-kde"><title>KDE</title>
 
-<para>The KDE Frameworks are a set of libraries for Qt 5 which form the basis of the Plasma desktop environment and the KDE Applications suite. Packaging a Frameworks-based library does not require any steps beyond those described above for general Qt-based libraries. Frameworks-based applications should not use <literal>makeQtWrapper</literal>; instead, use <literal>kdeWrapper</literal> to creat ...
+<para>The KDE Frameworks are a set of libraries for Qt 5 which form the basis of the Plasma desktop environment and the KDE Applications suite. Packaging a Frameworks-based package does not require any steps beyond those described above for general Qt-based packages.</para>
```
The manual should also mention, that `nix-build` will not work.

ttuegel committed:
qt58: determine plugin and import paths from PATH

Plugin and QML import paths were previously determined by NIX_PROFILES. Using
PATH instead allows Qt applications to work under nix-shell without further
modification.

ttuegel committed:
qt56: determine plugin and import paths from PATH

Plugin and QML import paths were previously determined by NIX_PROFILES. Using
PATH instead allows Qt applications to work under nix-shell without further modification.

gleber commented:
@ttuegel Looks like packages inited in https://github.com/NixOS/nixpkgs/pull/26524 were not covered by this PR. My understanding of changes in this PR is not good enough to fix it. Can you take a look?

Mic92 commented:
ultrastar-creator fixed in 558b5fbd534d7a9d68938ce304e0c97aaafa90ec

gleber commented:
@Mic92 ultrastar-manager still has references to `makeQtWrapper`

Mic92 commented:
9f14594cbc7313f6aac5ad58c1f789ad45ac6f24

gleber commented:
@Mic92 Awesome, you rock!

vcunat commented:
Any immediate ideas about this seemingly non-sensical error of kdiff3? http://hydra.nixos.org/build/55590912
> Could NOT find KF5 (missing: Crash) (found suitable version "5.34.0", minimum required is "5.16.0")

Bisection on staging narrowed-down the range into this PR's commits; bad: fc72335570, good: 6beea32ab1.  Interestingly, it does build on master.  There's also a similar error for kdevplatform: http://hydra.nixos.org/build/55596900

FRidh commented:
@vcunat I've been adding the missing dependencies to the packages. Maybe some dependencies were not propagated anymore.

vcunat commented:
Thanks, the related problems I know of are fixed on master, after 43ca91106 and a couple preceding commits.
